### PR TITLE
Partial updates on Ansible, Puppet and Chef

### PR DESCRIPTION
**Pull request by rambleraptor**
This has been too long coming.

How does Terraform do integration tests on updates? I can do a manual test to verify that everything doesn't crash + burn, but I'd like something more automated in the future.

-----------------------------------------------------------------
# [all]
Partial updates on A/C/P
## [puppet]
Partial updates
## [chef]
Partial updates
## [ansible]
Partial updates

### CODE REVIEW

**Comment by SirGitsalot** [comment]
Is there any easy way to see the generated code for a PR?

**Comment by rambleraptor** [comment]
After the magician runs, it'll give a list of PRs with the generated code. If you look up a couple comments, you'll see the Magician message with a link to [modular-magician/ansible#63](https://github.com/modular-magician/ansible/pull/63)

**Comment by rambleraptor** [comment]
I added in Puppet support for partial updates too.

I started on Chef, but it seems like it'll be a lot trickier because we don't have an easy way to do property diffs.

**Comment by danawillow** [comment]
Here's an example of a Terraform test that tests update: https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_compute_disk_test.go#L247. The Terraform libraries give us the ability to have multiple `TestStep`s in a given test, so in this case, it runs the basic config (https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_compute_disk_test.go#L601) and then runs terraform as if you had changed the config to the updated one (https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_compute_disk_test.go#L629). We don't have any way in Terraform to test that it was indeed an update and not a destroy/create, but this at least lets us check that the resource has the properties we expect at the end of the test.

File: templates/puppetchef/update_props.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<% update_props = properties_by_custom_update(object.all_user_properties) -%>
```
Comment (rambleraptor) [inline templates/puppetchef/update_props.erb]: I'd like to start putting the shared Puppet+Chef code inside a `puppetchef` folder to clean up the templates/ directory. Any objections?

File: api/type.rb [hunk 2]
```diff
@@ -200,9 +200,18 @@ def initialize(name)
...
+
+      def api_name
+        name
+      end
+    end
+
     # Represents a fingerprint.  A fingerprint is an output-only
     # field used for optimistic locking during updates.
-    class Fingerprint < String
+    class Fingerprint < FetchedExternal
```
Comment (ndmckinley) [inline api/type.rb]: Hm, what does that do to the Terraform output code?

File: templates/puppetchef/update_props.erb (hunk 1 above)
Comment (ndmckinley) [reply templates/puppetchef/update_props.erb]: SGTM.

File: api/type.rb (hunk 2 above)
Comment (rambleraptor) [reply api/type.rb]: This shouldn't do anything.

**Comment by rambleraptor** [comment]
@danawillow @ndmckinley @SirGitsalot ping!

sorry this took so long! Completely forgot about it.

File: api/type.rb (hunk 2 above)
Comment (ndmckinley) [reply api/type.rb]: Great!

File: api/type.rb [hunk 3]
```diff
@@ -166,15 +166,23 @@ def initialize(name)
...
       def validate
         super
         @output = true if @output.nil?
         # TODO(ndmckinley): This doesn't work in puppet, chef, or ansible.
```
Comment (danawillow) [inline api/type.rb]: If you're removing the line of code this comment refers to, can you also remove the comment?

File: products/compute/api.yaml [hunk 4]
```diff
@@ -146,7 +146,6 @@ objects:
           internally during updates.
         update_url: 'projects/{{project}}/regions/{{region}}/addresses/{{name}}/setLabels'
         update_verb: :POST
-        min_version: beta
```
Comment (danawillow) [inline products/compute/api.yaml]: I don't see labels in the GA api, do you? https://cloud.google.com/compute/docs/reference/rest/v1/addresses

**Comment by rambleraptor** [comment]
@ndmckinley pining you before magician run since you wanted to have this for release soon.

My testing strategy:
- Build disks on Puppet, Chef, Ansible
- Change that disk's size + labels (size is a good "regular" case, labels have fingerprints and are weird)
- Verify changes on GCP
- Delete disks.

I want to get full update tests written for all of these, but those will have to wait for the moment. My intent is that if disk size + labels work, everything else is one of those cases.

**Comment by ndmckinley** [comment]
I think you're probably right about the tests!  Sounds pretty good to me.

**Comment by ndmckinley** [comment]
Can you ping me again when you're ready with this one?  :)

**Comment by rambleraptor** [comment]
I'm logging off in a moment, but I just sent it to the Magician.

**Comment by rambleraptor** [comment]
@ndmckinley ping!

File: templates/ansible/resource.erb [hunk 5]
```diff
@@ -257,12 +266,54 @@ def main():
...
+    return <%= method %>
 <%   end # if object.input -%>
 <% else # object.update.nil? -%>
 <%= lines(indent(object.update, 4)) -%>
 <% end # object.update.nil? -%>
 
 
+<% unless update_props.empty? -%>
+def update_fields(module, request, response):
+<% empty = update_props.values.flatten.reject { |p| p.is_a? Api::Type::FetchedExternal }.empty? -%>
```
Comment (ndmckinley) [inline templates/ansible/resource.erb]: There's some rules in GOVERNANCE.MD about how to indent this stuff ... I'm not too bothered, honestly, but wanted to let you know.  Dana recently changed `terraform/resource.erb` to follow those rules and it looks pretty good.

**Comment by ndmckinley** [comment]
Woop woop!  This is huge.  :)

**Pull request merged**